File: testing.py
"""
Demonstration of the GazeTracking library.
Check the README.md for complete documentation.
"""

import cv2
from gaze_tracking import GazeTracking

gaze = GazeTracking()
webcam = cv2.VideoCapture(0)

# Screen dimensions
screen_width = 640
screen_height = 480

# Box dimensions for a 3x3 grid
box_width = screen_width // 3
box_height = screen_height // 3

while True:
    # We get a new frame from the webcam
    _, frame = webcam.read()

    # We send this frame to GazeTracking to analyze it
    gaze.refresh(frame)

    frame = gaze.annotated_frame()
    text = ""

    if gaze.is_blinking():
        text = "Blinking"
    elif gaze.is_right():
        text = "Looking right"
    elif gaze.is_left():
        text = "Looking left"
    elif gaze.is_center():
        text = "Looking center"

    cv2.putText(frame, text, (90, 60), cv2.FONT_HERSHEY_DUPLEX, 1.6, (147, 58, 31), 2)

    left_pupil = gaze.pupil_left_coords()
    right_pupil = gaze.pupil_right_coords()
    cv2.putText(frame, "Left pupil:  " + str(left_pupil), (90, 130), cv2.FONT_HERSHEY_DUPLEX, 0.9, (147, 58, 31), 1)
    cv2.putText(frame, "Right pupil: " + str(right_pupil), (90, 165), cv2.FONT_HERSHEY_DUPLEX, 0.9, (147, 58, 31), 1)

    # Get the gaze ratios
    horizontal_ratio = gaze.horizontal_ratio()
    vertical_ratio = gaze.vertical_ratio()

    # Create a visible grid of 9 boxes on the screen
    for row in range(3):
        for col in range(3):
            # Coordinates of the top-left and bottom-right corners of each box
            top_left = (col * box_width, row * box_height)
            bottom_right = ((col + 1) * box_width, (row + 1) * box_height)

            # Default color of the grid boxes
            box_color = (200, 200, 200)

            cv2.rectangle(frame, top_left, bottom_right, box_color, 2)

    if horizontal_ratio is not None and vertical_ratio is not None:

        # Draw the gaze location on the screen


         # Calculate the row and column based on the gaze ratios
        gaze_col = int(horizontal_ratio * 3) # 0, 1, or 2 for right, center, left
        gaze_row = int(vertical_ratio * 3)    # 0, 1, or 2 for top, center, bottom

        print(gaze_col, gaze_row)

        # If the current box matches the gaze location, change its color
        top_left = (gaze_col * box_width, gaze_row * box_height)
        bottom_right = ((gaze_col + 1) * box_width, (gaze_row + 1) * box_height)

        box_color = (0, 255, 0)  # Green color for the box being looked at

        # Draw the rectangle for the grid box
        cv2.rectangle(frame, top_left, bottom_right, box_color, 2)

        # Additionally, draw a small circle at the exact gaze point for better visualization
        gaze_x = int(horizontal_ratio * screen_width)
        gaze_y = int(vertical_ratio * screen_height)
        cv2.circle(frame, (gaze_x, gaze_y), 5, (0, 0, 255), -1)  # Red circle for gaze point

    cv2.imshow("Demo", frame)

    if cv2.waitKey(1) == 27:
        break
   
webcam.release()
cv2.destroyAllWindows()

File: testing.js
const cv = require('opencv4nodejs');
const GazeTracking = require('./gazeTracking'); // Make sure the correct path is set for your GazeTracking class

const gaze = new GazeTracking();
const webcam = new cv.VideoCapture(0);

// Screen dimensions
const screenWidth = 640;
const screenHeight = 480;

// Box dimensions for a 3x3 grid
const boxWidth = Math.floor(screenWidth / 3);
const boxHeight = Math.floor(screenHeight / 3);

// Function to put text on the frame
function putText(frame, text, x, y) {
    const point = new cv.Point(x, y);
    frame.putText(text, point, cv.FONT_HERSHEY_DUPLEX, 1.6, new cv.Vec(147, 58, 31), 2);
}

function main() {
    setInterval(() => {
        // We get a new frame from the webcam
        let frame = webcam.read();

        // We send this frame to GazeTracking to analyze it
        gaze.refresh(frame);

        frame = gaze.annotatedFrame();
        let text = "";

        if (gaze.isBlinking()) {
            text = "Blinking";
        } else if (gaze.isRight()) {
            text = "Looking right";
        } else if (gaze.isLeft()) {
            text = "Looking left";
        } else if (gaze.isCenter()) {
            text = "Looking center";
        }

        putText(frame, text, 90, 60);

        const leftPupil = gaze.pupilLeftCoords();
        const rightPupil = gaze.pupilRightCoords();
        putText(frame, `Left pupil:  ${leftPupil}`, 90, 130);
        putText(frame, `Right pupil: ${rightPupil}`, 90, 165);

        // Get the gaze ratios
        const horizontalRatio = gaze.horizontalRatio();
        const verticalRatio = gaze.verticalRatio();

        // Create a visible grid of 9 boxes on the screen
        for (let row = 0; row < 3; row++) {
            for (let col = 0; col < 3; col++) {
                // Coordinates of the top-left and bottom-right corners of each box
                const topLeft = new cv.Point(col * boxWidth, row * boxHeight);
                const bottomRight = new cv.Point((col + 1) * boxWidth, (row + 1) * boxHeight);

                // Default color of the grid boxes
                const boxColor = new cv.Vec(200, 200, 200);
                frame.drawRectangle(topLeft, bottomRight, boxColor, 2);
            }
        }

        if (horizontalRatio !== null && verticalRatio !== null) {
            // Calculate the row and column based on the gaze ratios
            const gazeCol = Math.floor(horizontalRatio * 3); // 0, 1, or 2 for right, center, left
            const gazeRow = Math.floor(verticalRatio * 3);    // 0, 1, or 2 for top, center, bottom

            // If the current box matches the gaze location, change its color
            const topLeft = new cv.Point(gazeCol * boxWidth, gazeRow * boxHeight);
            const bottomRight = new cv.Point((gazeCol + 1) * boxWidth, (gazeRow + 1) * boxHeight);
            const boxColor = new cv.Vec(0, 255, 0);  // Green color for the box being looked at

            // Draw the rectangle for the grid box
            frame.drawRectangle(topLeft, bottomRight, boxColor, 2);

            // Additionally, draw a small circle at the exact gaze point for better visualization
            const gazeX = Math.floor(horizontalRatio * screenWidth);
            const gazeY = Math.floor(verticalRatio * screenHeight);
            frame.drawCircle(new cv.Point(gazeX, gazeY), 5, new cv.Vec(0, 0, 255), -1);  // Red circle for gaze point
        }

        cv.imshow("Demo", frame);
        if (cv.waitKey(1) === 27) { // ESC key to break
            clearInterval();
            webcam.release();
            cv.destroyAllWindows();
        }
    }, 0);
}

main();
